Avoid compiler crash with missing transitive dependencies

retronym opened the pull request:
When we unpickle a reference to `a.b.C` and we don't have a package symbol for `a.b`, we synthesize a stub symbol. This code was modifed in b47aaf6445a to mutate the flags of a stub symbol rather than trying to create a particular subclass of `StubSymbol` when we figure out that a stub represents a package. However, this neglected to set the `MODULE` flag, which can lead to an assertion error later in the compiler.

This PR includes a direct fix for this flag error.

It also includes a few improvements to avoid forcing symbol info transformers in the compiler backend that I identified during the investigation.
I have a few more changes of this nature in [another branch](https://github.com/scala/scala/compare/2.12.x...retronym:topic/stubby-3?expand=1) that I'll revisit.

retronym commented:
TODO:
 - [ ] Does this fix [SI-10171](https://issues.scala-lang.org/browse/SI-10171) ?

retronym commented:
@lrytz Could you please take a look at these commits to see if they are on the right track?

The first one avoids tripping an assertion with a stub symbol during symbol substitution that was happening in Uncurry's info transformer, which had dealiased a method parameter from `Predef.String` to `java.lang.String`, but crashed because the method return type was a stub. Special casing `boundSyms` avoids the problem there.

The second one independentally fixes the problem by stopping the propagation of info transformers sooner, by disabling the specialization transform when the compiler has reached the backend. I believe this is safe, even for multi-run compilation (in that case we discard all but the first element of the info history on the next run).

The specialization info transform is the worst offender in forcing method infos. We could potentially disable parts of other info transforms under the same condition, e.g. the [addition of outer accessor methods](https://github.com/scala/scala/blob/d2d33ddf8ce4dff503c8830958bac0e449eb7d34/src/compiler/scala/tools/nsc/transform/ExplicitOuter.scala#L178-L190) in `ExplicitOuter`.

The final commit attempts to lazify part of the conversion from Symbols to BTypes. I haven't validated whether or how much this actually limits the traversal of symbol infos.

A test project is attached to https://issues.scala-lang.org/browse/SI-10171, I'm going to try to extract a `run` test from that.

retronym committed:
Avoid stub symbol related crash in backend

In this test case, the backend forces the specialization
info transform of `Sub` during computation of its
inner class metadata. This in turn runs the info transforms
of the `Base`. This leads to the uncurry info tranform
transforming a signature that has a type alias as a
method parameter type. Subsequent substution of the new
method symbol into the result type, which includes a
stub symbol for an absent class, tripped an assertion:

```
requirement failed: package b
java.lang.IllegalArgumentException: requirement failed: package b
	at scala.Predef$.require(Predef.scala:277)
	at scala.reflect.internal.Types$ModuleTypeRef.<init>(Types.scala:1879)
	at scala.reflect.internal.Types$PackageTypeRef.<init>(Types.scala:1897)
	at scala.reflect.internal.Types$TypeRef$.apply(Types.scala:2401)
	at scala.reflect.internal.Types.typeRef(Types.scala:3553)
	at scala.reflect.internal.Types.typeRef$(Types.scala:3536)
	at scala.reflect.internal.SymbolTable.typeRef(SymbolTable.scala:16)
	at scala.reflect.internal.Symbols$TypeSymbol.newTypeRef(Symbols.scala:3026)
	at scala.reflect.internal.Symbols$TypeSymbol.updateTypeCache(Symbols.scala:3079)
	at scala.reflect.internal.Symbols$TypeSymbol.maybeUpdateTypeCache(Symbols.scala:3065)
	at scala.reflect.internal.Symbols$TypeSymbol.tpe_$times(Symbols.scala:3043)
	at scala.reflect.internal.Symbols$Symbol.typeOfThis(Symbols.scala:2020)
	at scala.reflect.internal.Types$ThisType.underlying(Types.scala:1184)
	at scala.reflect.internal.Types$SimpleTypeProxy.boundSyms(Types.scala:150)
	at scala.reflect.internal.Types$SimpleTypeProxy.boundSyms$(Types.scala:150)
	at scala.reflect.internal.Types$SingletonType.boundSyms(Types.scala:1088)
	at scala.reflect.internal.tpe.TypeMaps$SubstMap.apply(TypeMaps.scala:726)
	at scala.reflect.internal.tpe.TypeMaps$SubstSymMap.apply(TypeMaps.scala:789)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:102)
	at scala.reflect.internal.tpe.TypeMaps$SubstSymMap.apply(TypeMaps.scala:783)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:102)
	at scala.reflect.internal.tpe.TypeMaps$SubstSymMap.apply(TypeMaps.scala:783)
	at scala.reflect.internal.Types$Type.substSym(Types.scala:727)
	at scala.reflect.internal.tpe.TypeMaps$TypeMap.mapOver(TypeMaps.scala:123)
	at scala.reflect.internal.transform.UnCurry$$anon$1.apply(UnCurry.scala:53)
	at scala.reflect.internal.transform.UnCurry.transformInfo(UnCurry.scala:154)
```

This commit address the direct failure above by setting coherent flags on
the stub package class symbol (it also needs the MODULE flag).

retronym commented:
@lrytz I've added a test case and another simple fix (give the stub symbol the `MODULE` flag). However, I think the other commits still have value: avoiding needless work limits our surface area to future bugs and helps performance.

retronym commented:
/cc @jvican

retronym commented:
/rebuild

retronym commented:
/rebuild

retronym commented on src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:
```diff
@@ -1199,7 +1199,7 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
    *  If it is a 'no-specialization' run, it is applied only to loaded symbols.
    */
   override def transformInfo(sym: Symbol, tpe: Type): Type = {
-    if (settings.nospecialization && currentRun.compiles(sym)) tpe
```
The backend needs to specialize `TupleN` to populate its list of tuple constructors to make tests for box-elimination to work. This seems to be the only case covered by our test suite that requires specialization of a type after the specialization phase itself has run.

retronym committed:
Avoid forcing info transforms of primitive methods

Although this is cheap, when debugging log output of info
transformer activity this was a major source of noise.

This commit avoids the info lookup for methods other
than `+`, and then for `+` uses the typer phase info
to distinguish concatentation from addition.

lrytz commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -1194,6 +1194,7 @@ trait Types
       else super.safeToString
     override def narrow: Type = this
     override def kind = "ThisType"
+    override def boundSyms = if (sym.isInstanceOf[StubSymbol]) emptySymbolSet else super.boundSyms
```
Looks OK, but looking around for a while, it seems that allowing to access `tpe` on stub symbols is a feature. So I prefer the fix in your next commit (`MODULE` flag for stub package symbols), and we could live without the special-case here.

```
scala> import scala.reflect.internal.Flags._
scala> val c = definitions.ScalaPackageClass.newStubSymbol(newTypeName("C"), "not here")
scala> c.tpe
res7: $r.intp.global.Type = C

scala> val c = definitions.ScalaPackageClass.newStubSymbol(newTypeName("C"), "not here")
scala> c.setFlag(PACKAGE)
scala> c.tpe
java.lang.IllegalArgumentException: requirement failed: package C
  at scala.reflect.internal.Types$ModuleTypeRef.<init>(Types.scala:1879)

scala> val c = definitions.ScalaPackageClass.newStubSymbol(newTypeName("C"), "not here")
scala> c.setFlag(PACKAGE | MODULE)
scala> c.tpe
res12: $r.intp.global.Type = C.type
```

While looking at `boundSyms`, do you know why [`ExistentialType.boundSyms`](https://github.com/scala/scala/blob/v2.12.1/src/reflect/scala/reflect/internal/Types.scala#L2619) doesn't `++ underlying.boundSyms`?

retronym replied:
> While looking at `boundSyms`, do you know why `ExistentialType.boundSyms` doesn't `++ underlying.boundSyms`?

Looks like it is called repeatedly as the [substituition `TypeMap`](https://github.com/scala/scala/blob/1eb331b4196492f8f8084a2c102272c3a9c0cd2d/src/reflect/scala/reflect/internal/tpe/TypeMaps.scala#L696-L759) recurses, so its doing the right thing by being shallow. 

Come to think of it, I suspect we could actually just return null (in `SimpleTypeProxy.boundSyms`) and have the same results as we do today.

lrytz replied:
Would you like to keep this override, given the other more fundamental fix? I'd remove it, but either way is fine.

lrytz replied:
just a quick ping here, then we can merge.

adriaanm replied:
Does this mean the review is currently still pending? Lets make that clear with "request changes" review. I was about to merge this PR because the review status was "approved".

lrytz replied:
ok, yes, i just wanted @retronym to consider this once more.

retronym replied:
I've removed the override for now. I think we could actually generalize it to:

```scala
class SubType {
  override def boundSyms = emptySymbolSet
}
```

Along with overrides to pick that:

```
abstract case class ThisType {
  override def boundSyms = super[SubType].boundSyms
}
...
```

But let's rely on `underlying.boundSyms` for now.

lrytz commented on src/reflect/scala/reflect/internal/pickling/UnPickler.scala:
```diff
@@ -377,7 +377,7 @@ abstract class UnPickler {
       def readThisType(): Type = {
         val sym = readSymbolRef() match {
-          case stub: StubSymbol => stub.setFlag(PACKAGE)
+          case stub: StubSymbol => stub.setFlag(PACKAGE | MODULE)
```
This looks good.

lrytz commented on src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:
```diff
@@ -1199,22 +1204,34 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
+          assert(!currentRun.compiles(sym), sym)
+          // Skip specialization info transform for third party classes that aren't referenced directly
+          // from the tree or by the specialization info transform itself that are run up to the end of
+          // the specialization phase.
```
this comment is a bit hard to read

lrytz commented on src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:
```diff
@@ -198,6 +198,11 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
+    override def apply(unit: CompilationUnit): Unit = {
+      super.apply(unit)
+      FunctionClass.seq.map(_.info)
+      TupleClass.seq.map(_.info)
```
there should be a comment here

lrytz reviewed (commented):
This looks good! Also the commit introducing lazy completion, though I only took a quick look.

retronym commented:
@lrytz I've trimmed this down to the essential commit and some low-risk passengers.

dwijnand commented:
super green

lrytz reviewed (changes requested):
()

retronym committed:
Remove non-essential fix for stub symbol failure

Given that we correctly setup the flags on the stub symbol,
we no longer trip an assertion in ModuleTypeRef's constructor.
